### PR TITLE
Misc fixes

### PR DESCRIPTION
**Pull request by selvanair**
This completes all "bugs" I found while looking for what causes #577. None of these would explain the crash in #577 though. I also have a fix for #576, but it requires more testing. For lack of time, leaving it for 2.6.1.

### CODE REVIEW

**Comment by selvanair** [comment]
strawberry-perl is a real pita.. And, all these repeated openssl builds are anyway a huge waste of time and resources. All for a "password-change" feature of limited use that works only for private key files on disk, not inline.

Shall we get rid of it, or replace with a script that would depend on openssl.exe? We could install the latter by default.

**Comment by shipitsin-ilia** [comment]
changing password for on disk keys is considered as somewhat useful, however, I'm not sure that it belongs to the scope of vpn software.

also, I found that inlined key (embedded into config file) cannot be enctypted that way.

I vote for getting rid of it completely, there are many tools for encrypting private keys anyway.

**Comment by cron2** [comment]
Hi,

On Thu, Jan 19, 2023 at 09:37:00PM -0800, Selva Nair wrote:
> strawberry-perl is a real pita.. And, all these repeated openssl builds are anyway a huge waste of time and resources. All for a "password-change" feature of limited use that works only for private key files on disk, not inline.
> 
> Shall we get rid of it, or replace with a script that would depend on openssl.exe? We could install the latter by default.

If that is really the only use ("change private key pass phrase for a
non-embedded .pem file on disk") I say get rid of it, and replace with
a help popup that shows a sample command line.

Lots of complication for a rarely-used feature.

Having a good mechanism to such keys into windows key store and then 
just throw away the original file would be much more secure anyway...
(we do not have this, right?)

gert

-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             ***@***.***

**Comment by chipitsine** [comment]
пт, 20 янв. 2023 г. в 13:53, Gert Doering ***@***.***>:

> Hi,
>
> On Thu, Jan 19, 2023 at 09:37:00PM -0800, Selva Nair wrote:
> > strawberry-perl is a real pita.. And, all these repeated openssl builds
> are anyway a huge waste of time and resources. All for a "password-change"
> feature of limited use that works only for private key files on disk, not
> inline.
> >
> > Shall we get rid of it, or replace with a script that would depend on
> openssl.exe? We could install the latter by default.
>
> If that is really the only use ("change private key pass phrase for a
> non-embedded .pem file on disk") I say get rid of it, and replace with
> a help popup that shows a sample command line.
>
> Lots of complication for a rarely-used feature.
>
> Having a good mechanism to such keys into windows key store and then
> just throw away the original file would be much more secure anyway...
> (we do not have this, right?)
>

those private keys are usually generated on openvpn server side and
distributed to user as config/embedded or zip archive with several files.

I'm afraid that "generate private key in windows key store and let openvpn
server to sign that key"  is never used in real world.

**Comment by cron2** [comment]
Hi,

On Fri, Jan 20, 2023 at 12:10:35AM -0800, Ilya Shipitsin wrote:
> > Having a good mechanism to such keys into windows key store and then
> > just throw away the original file would be much more secure anyway...
> > (we do not have this, right?)
> 
> those private keys are usually generated on openvpn server side and
> distributed to user as config/embedded or zip archive with several files.
> 
> I'm afraid that "generate private key in windows key store and let openvpn
> server to sign that key"  is never used in real world.

The sentence above was missing an "... to *import* such keys...", so 
no matter where the keys come from, if we can import them into a secure
key storage and then remove the file, it would be nice.

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             ***@***.***

File: pkcs11.c [hunk 1]
```diff
@@ -663,7 +663,7 @@ QueryPkcs11DialogProc(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam)
             {
                 DeleteObject(hf);
             }
-            RemoveProp(hwndDlg, cfgProp);
+            RemoveProp(hwndDlg, L"header_font");
```
Comment (lstipakov) [inline pkcs11.c:665]: For `cfgProp` we use `const TCHAR *cfgProp = _T("conn");`, shouldn't we do the same for `header_font`?

**Comment by selvanair** [comment]
Hi

On Fri, Jan 20, 2023 at 3:21 AM Gert Doering ***@***.***>
wrote:

> Hi,
>
> On Fri, Jan 20, 2023 at 12:10:35AM -0800, Ilya Shipitsin wrote:
> > > Having a good mechanism to such keys into windows key store and then
> > > just throw away the original file would be much more secure anyway...
> > > (we do not have this, right?)
> >
> > those private keys are usually generated on openvpn server side and
> > distributed to user as config/embedded or zip archive with several files.
> >
> > I'm afraid that "generate private key in windows key store and let
> openvpn
> > server to sign that key" is never used in real world.
>
> The sentence above was missing an "... to *import* such keys...", so
> no matter where the keys come from, if we can import them into a secure
> key storage and then remove the file, it would be nice.
>

"Import" and then use cryptoapicert is what I do with almost all Windows
users I deal with. Very easy with pkcs12 files -- just double click and an
import dialog opens up. PEM certificate files can also be imported
similarly, but AFAIK, there is no easy way to import into Windows PEM key
files that OpenSSL generates (encrypted or not).

Selva

**Review by lstipakov — APPROVED** [review]
Multiple settings dialog is something I was also annoyed by. Slightly tested, looks good.

**Comment by chipitsine** [comment]
пт, 20 янв. 2023 г. в 14:21, Gert Doering ***@***.***>:

> Hi,
>
> On Fri, Jan 20, 2023 at 12:10:35AM -0800, Ilya Shipitsin wrote:
> > > Having a good mechanism to such keys into windows key store and then
> > > just throw away the original file would be much more secure anyway...
> > > (we do not have this, right?)
> >
> > those private keys are usually generated on openvpn server side and
> > distributed to user as config/embedded or zip archive with several files.
> >
> > I'm afraid that "generate private key in windows key store and let
> openvpn
> > server to sign that key" is never used in real world.
>
> The sentence above was missing an "... to *import* such keys...", so
> no matter where the keys come from, if we can import them into a secure
> key storage and then remove the file, it would be nice.
>

the sentence is not missing anything ))

yes, an average user can import p12 and import openvpn config, but it is
near to impossible to teach user to adopt
openvpn config to use imported file

however, such a feature can be really nice

**Comment by cron2** [comment]
Hi,

On Fri, Jan 20, 2023 at 01:09:25AM -0800, Selva Nair wrote:
> "Import" and then use cryptoapicert is what I do with almost all Windows
> users I deal with. Very easy with pkcs12 files -- just double click and an
> import dialog opens up. PEM certificate files can also be imported
> similarly, but AFAIK, there is no easy way to import into Windows PEM key
> files that OpenSSL generates (encrypted or not).

This is somewhat outside the realms of *this* ticket... but, this is 
good to know.  So I might change to "tell my windows users to just import
the pkcs12 files they get from the inhouse CA anyway, and have OpenVPN
reference it" (--cryptoapicert "SUBJ:$myname").

Is there a good trick to get something generic into cryptoapicert,
like the CA name ("when connecting to $customer, use any non-expired
cert in the CA store, but only those from $customerca") so users do
not have to modify the .ovpn file?

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             ***@***.***

**Comment by selvanair** [comment]
> near to impossible to teach user to adopt
> openvpn config to use imported file

What I do is to distribute configs with `cryptoapi-cert THUMB:<thumbprint>` already filled in,  and a p12 file that the user has to just double click and import. The thumbprint is known as soon as the certificate is generated. For really clueless users I provide a script that does this.

That said, improving our import feature to handle multiple files (dependencies) and do this on the fly is a nice project for those anxious to contribute :)

File: pkcs11.c (hunk 1 above)
Comment (selvanair) [reply pkcs11.c:665]: We can but its only used once so far...

**Comment by selvanair** [comment]
> Is there a good trick to get something generic into cryptoapicert,
> like the CA name ("when connecting to $customer, use any non-expired
> cert in the CA store, but only those from $customerca") so users do
> not have to modify the .ovpn file?

Not currently. I find SUBJ: error-prone/ambiguous, so use thumb, but that requires per-user config generation. There was a patch to add a selector based on "templates" but that's a Windows CA thing and not much use for most of us. 

I think we can easily add a CA-based selection though its also as ambiguous (or even more) as SUBJ:. But could work for some.

**Comment by chipitsine** [comment]
пт, 20 янв. 2023 г. в 15:20, Selva Nair ***@***.***>:

> near to impossible to teach user to adopt
> openvpn config to use imported file
>
> What I do is to distribute configs with cryptoapi-cert THUMB:<thumbprint>
> already filled in, and a p12 file that the user has to just double click
> and import. The thumbprint is known as soon as the certificate is
> generated. For really clueless users I provide a script that does this.
>

personally, I am interested in using computer self issued certificates for
PLAP.
but haven't found spare time yet ((

File: pkcs11.c (hunk 1 above)
Comment (lstipakov) [reply pkcs11.c:665]:
```
SetProp(parent, L"header_font", (HANDLE)hfb);	C:\Users\lev\Projects\openvpn-build\src\openvpn-gui\pkcs11.c	419	31
HFONT hf = (HFONT) GetProp(hwndDlg, L"header_font");	C:\Users\lev\Projects\openvpn-build\src\openvpn-gui\pkcs11.c	660	51
RemoveProp(hwndDlg, L"header_font");	C:\Users\lev\Projects\openvpn-build\src\openvpn-gui\pkcs11.c	665	35
```

Comment (selvanair) [reply pkcs11.c:665]: I know, "once" was used in plural :) (or all in one dialog)...  Well, I'm convinced: will define a variable.

**Comment by lstipakov** [comment]
> strawberry-perl is a real pita.. And, all these repeated openssl builds are anyway a huge waste of time and resources. 

I very much agree.  In openvpn-build (which is triggered on every commit to openvpn) we build openssl twice (because of vcpkg triplet thing) - for gui and for openvpn. This makes everything slow.

**Pull request merged**